keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,1069 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+If we hard coded the `UID` and `GID`, we could end up in a scenario where those are in use by another process on the machine, which would expose some of the credentials accessible to the `UID` and `GID`s to that process. So we plan to use adduser --system or using the appropriate ranges from /etc/login.defs instead of hard coding the `UID` and `GID`.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In `kubeadm` the control-plane components are run as [static-pods](https://kubernetes.io/docs/concepts/workloads/pods/#static-pods), i.e. pods directly managed by kubelet. We can use the `runAsUser` and `runAsGroup` fields in [SecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) to run the containers in these Pods with a unique `UID` and `GID` and the `capabilities` field to drop all capabilities other than the ones required.
+
+### Assigning UID and GID
+There are 3 options for setting the `UID`/`GID` of the control-plane components:-
+
+1. **Update the kubeadm API to make the uid/gid configurable by the user:**  This can be implemented in 2 ways:-
+    1. This can be done by adding fields `UID` and `GID` of type `int64` to the `ControlPlaneComponent` `struct` in https://github.com/kubernetes/kubernetes/blob/854c2cc79f11cfb46499454e7717a86d3214e6b0/cmd/kubeadm/app/apis/kubeadm/types.go#L132 as demonstrated in [this](https://github.com/kubernetes/kubernetes/pull/99753) PR.
+    2. User provides `kubeadm` with a range of uid/gids that are safe to use i.e. are not being used by another user.
+
+2. **Use constant values for uid/gid:** The `UID` and `GID`for each of the control-plane components would be set to some predetermined value in https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go.
+
+3. **Create system users and let users override the defaults:** Use `adduser --system` or equivalent to create `UID`s in the SYS_UID_MIN - SYS_UID_MAX range and groupadd --system or equivalent to create `GID`s in the SYS_GID_MIN - SYS_GID_MAX range. Additionally if users want to specify their own `UID`s or `GID`s we will support that through `kubeadm` v1beta3 API.
```

Comment:
We could offset the UID and GID used here with some really high value as SYS_UID_MIN to prevent chances of overlap. However, for some flavors of kubernetes this may become a problem as they have their own way of assigning unique User IDs for running pods e.g. openshift: https://www.openshift.com/blog/a-guide-to-openshift-and-uids

Reply:
My 2 cents would be to keep this logic simple as to fit everyones requirements by default would be extremely challenging. Giving them the ability to pick their own ids through the API would solve the problem on these flavors.

Reply:
You are right. I am definitely in favor of simple and secure.